app/livechat/server/api/v1/room.js

```diff
@@ -179,3 +182,54 @@ API.v1.addRoute('livechat/room.forward', { authRequired: true }, {
 		API.v1.success(Meteor.runAsUser(this.userId, () => Meteor.call('livechat:transfer', this.bodyParams)));
 	},
 });
+
+API.v1.addRoute('livechat/room.visitor', { authRequired: true }, {
+	put() {
+		try {
+			check(this.bodyParams, {
+				rid: String,
+				oldVisitorId: String,
+				newVisitorId: String,
+			});
+
+			const user = Promise.await(Users.findOneById(this.userId));
+			if (!user) {
+				throw new Error('error-user-not-found');
+			}
+
+			if (!Promise.await(hasPermissionAsync(this.userId, 'change-livechat-room-visitor'))) {
+				throw new Error('error-not-authorized');
+			}
+
+			const { rid, visitorId } = this.bodyParams;
+
+			const { visitor } = Promise.await(findVisitorInfo({ userId: this.userId, visitorId }));
+			if (!visitor) {
+				throw new Meteor.Error('invalid-visitor-token');
+			}
+
+			const room = Promise.await(LivechatRooms.findOneById(rid));
+			if (!room) {
+				throw new Meteor.Error('invalid-room-id');
+			}
+
+			if (!canAccessRoom(room, user)) {
+				throw new Error('error-not-allowed');
+			}
+
+			Promise.await(LivechatRooms.changeVisitorByRoomId(rid, visitor));
```

Comment:
Create a method [here](https://github.com/RocketChat/Rocket.Chat/blob/develop/app/livechat/server/lib/Livechat.js) to replace the room visitor, then move the following sentences to the new method:
```
Promise.await(LivechatRooms.changeVisitorByRoomId(rid, visitor));
const { _id, username, token } = visitor;
Promise.await(Livechat.notifyRoomVisitorChange(rid, { _id, username, token }));
```

Reply:
Added this change in the latest commit